# Jupyter One
---
This is my first Jupyter project and will be a guided project from [Dataquest.io](https://www.dataquest.io/).

In this project, the goal is to assist a company that makes free apps to find the types of apps that attracts the most users. As the company's profit is based on advertisements, creating the most popular type of apps directly influence their income. Based on data from both the Apple's Apps Store and Google's Play store, this project will complete the following goals:

* Retrieve relevant data from each data set.
* Remove inaccurate data.
* Remove duplicate entries.
* Remove non-English data.
* Isolate free apps for analysis.

    

---
### Data Set Information
Of the millions of apps available on both platform, we are taking a small sample from each.

Apple Store's data set documentation can be found [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/version/7). This data set is a sample data from 7,197 apps found in Apple's App Store.

Google Play Store's data set documentation can be found [here](https://www.kaggle.com/lava18/google-play-store-apps/home). This data set is a sample data from 10,841 apps found in Google's Play Store.

In [35]:
from csv import reader

### App Store data
ios_file = open('.\dataset\AppleStore\AppleStore.csv', encoding='utf8')
read_ios = reader(ios_file)
ios = list(read_ios)
app_header = ios[0]
app = ios[1:]

android_file = open('.\dataset\PlayStore\googleplaystore.csv', encoding='utf8')
read_android = reader(android_file)
android = list(read_android)
play_header = android[0]
play = android[1:]

Before we start and to make things easier, the function `explore_data()` is created to go through a data set's information.

In [32]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

---
**Useful Columns**

App Store's columns that are useful for our goal are: `track_name`[2], `currency`[4], `price`[5], `rating_count_tot`[6], `rating_count_ver`[7], `prime_genre`[12].

Google Play Store's columns that are useful are: `App`[0], `Category`[1], `Reviews`[3], `Installs`[5], `Type`[6], `Price`[7], `Genre`[9].


In [33]:
print('App Store header\n', app_header)
explore_data(app, 0, 1, True)
print('\n')
print('Play Store header\n', play_header)
explore_data(play, 0, 1, True)


App Store header
 ['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


Number of rows: 7197
Number of columns: 17


Play Store header
 ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


---

### Data Clean Up

**Issue with data**

From [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion) of Google Play Store's documentation, an error was found on one of the row. From one of the [discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/81460), looks like the 'Category' data is missing and it cause this row's data to be shift left one column. We will omit this row from our Play Store's sample data.

In [4]:
print(play[10472])
del play[10472]
print('\n', play[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']

 ['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


---

**Duplicate Apps Entries**

I have separate the apps into a unique list and duplicate list. Looks like there are 1,181 duplicate entries in the Google Play Store's data set. Grabbing a random app from our duplicate list, the only difference between the duplicate is the number of reviews in column 4 (index 3). The app's data was retrieved at different times and the number of reviews have increased on the subsequence retrieval. I will use this difference find the most recent entry which should be the one with the most reviews. The other duplicates will be removed for each duplicate entry.

In [5]:
unique_list = []
duplicate_list = []

for row in play:
    name = row[0]
    if name in unique_list:
        duplicate_list.append(name)
    else:
        unique_list.append(name)

print('Number of unique Play Store apps: ', len(unique_list))
print('Number of duplicate Play Store apps: ', len(duplicate_list))
print('\n')

for row in play:
    dup_name = duplicate_list[1110]
    app_name = row[0]
    if dup_name == app_name:
        print(row)

Number of unique Play Store apps:  9659
Number of duplicate Play Store apps:  1181


["McDonald's", 'FOOD_AND_DRINK', '3.6', '145323', '42M', '10,000,000+', 'Free', '0', 'Everyone', 'Food & Drink', 'July 31, 2018', '5.15.0', '4.4 and up']
["McDonald's", 'FOOD_AND_DRINK', '3.6', '145646', '42M', '10,000,000+', 'Free', '0', 'Everyone', 'Food & Drink', 'July 31, 2018', '5.15.0', '4.4 and up']


---

**Duplicate Apps Entries**

I have separate the apps into a unique list and duplicate list. Looks like there are 1,181 duplicate entries in the Google Play Store's data set. Grabbing a random app from our duplicate list, the only difference between the duplicate is the number of reviews in column 4 (index 3). The app's data was retrieved at different times and the number of reviews have increased on the subsequence retrieval. I will use this difference find the most recent entry which should be the one with the most reviews. The other duplicates will be removed for each duplicate entry.

In [6]:
max_reviews = {}

for row in play:
    reviews = int(row[3])
    name = row[0]
    
    if name not in max_reviews:
        max_reviews[name] = reviews
    elif reviews > max_reviews[name]:
        max_reviews[name] = reviews
        
print('Expected length:', len(play) - 1181)
print('Length of dictionary:', len(max_reviews))
print("Expected reviews numbers for 'McDonald's: 145646")
print("Length in dictionary for 'McDonald's:", max_reviews["McDonald's"])

Expected length: 9659
Length of dictionary: 9659
Expected reviews numbers for 'McDonald's: 145646
Length in dictionary for 'McDonald's: 145646


---

**Removing Duplicates**

Using the dictionary `max_reviews` as a criteria, I loop through all entries in the original Google Play Store data set and only retrieve the entry with the max reviews number into the cleaned list `play_clean`. There are also entries that have the same max reviews number but we only want one. Thus, to in addition to adding the entry to the `play_clean` list, I will also tracked if I have already added the entry in another list called `already_added`. This will be another criterion in order to filter out duplicate entries that have the same max reviews number.

The expected unique entries in `play_clean` matches the actual value. Also, by running the code to find number of unique entries again, we can see that there are no duplicates entries.

In [7]:
play_clean = []
already_added = []

for row in play:
    reviews = int(row[3])
    name = row[0]
    
    if reviews == max_reviews[name] and name not in already_added:
        play_clean.append(row)
        already_added.append(name)

print('Expected length:', len(play) - 1181)
print('Length of cleaned Google Play Store list:', len(play_clean))        

unique_list = []
duplicate_list = []

for row in play_clean:
    name = row[0]
    if name in unique_list:
        duplicate_list.append(name)
    else:
        unique_list.append(name)

print('\nNumber of unique Play Store apps: ', len(unique_list))
print('Number of duplicate Play Store apps: ', len(duplicate_list))

Expected length: 9659
Length of cleaned Google Play Store list: 9659

Number of unique Play Store apps:  9659
Number of duplicate Play Store apps:  0


---

## Removing Non-English Apps

**Part 1**

From the samples above, we can see that there are some non-english apps in both Apple's App Store and Google's Play Store data sets. The English set of characters have **ASCII** values between 0-127. With that in mind, I will create a function `check_english_char()` that will take in a String (name of app) and will return `True` if all characters in the String is within the range of the English language's ASCII values. However, it will return `False` only if there are more than 3 non-English characters. This is in consideration for some apps with one or more ASCII values that are outside the range such as the trademark symbol (™) and icons(😜). This is not perfect, but it will do for now.

In [8]:
print(app[932][2])
print(app[934][2])
print(play_clean[4412][0])
print(play_clean[7940][0])

def check_english_char(a_string):
    num_of_non_english = 0
    
    for ch in a_string:
        unicode_num = ord(ch)
        if unicode_num > 127:
            num_of_non_english += 1
            if num_of_non_english > 3:
                return False
        
    return True

print('\n')
print(check_english_char('Instagram'))
print(check_english_char('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_english_char('Docs To Go™ Free Office Suite'))
print(check_english_char('Instachat 😜'))

腾讯视频-欢乐颂2全网首播
搜狐视频-欢乐颂2 全网首播
中国語 AQリスニング
لعبة تقدر تربح DZ


True
False
True
True


---

**Part 2**

After removing non-English apps, Apps Store list have **6183** apps remaining while Play Store list have **9614** apps left.

In [9]:
cleaner_app = []
cleaner_play = []

for row in app:
    name = row[2]
    
    if check_english_char(name):
        cleaner_app.append(row)
        
for row in play_clean:
    name = row[0]
    
    if check_english_char(name):
        cleaner_play.append(row)
        
print("After non-English removal, Apps Store list have {} entries.".format(len(cleaner_app)))
print("After non-English removal, Play Store list have {} entries.".format(len(cleaner_play)))

After non-English removal, Apps Store list have 6183 entries.
After non-English removal, Play Store list have 9614 entries.


---

**Isolating Free Apps **

**Part 1**

* App Store have **3222** free apps.
* Play Store have **8864** free apps.

Last step is to isolate all the apps that are free in each data set. As mentioned in the beginning, we are looking for apps that has the most users and what are the common qualities these types of apps may have. The steps we are taking to validate our strategy are:
1. Build an Android app and add it to Google Play Store.
2. If there are good response from users, we continue to develope the app further.
3. After 6 months and if the app is still profitable, we will also build an iOS version and add it to the Apple App Store.



In [10]:
cleanest_app = []
cleanest_play = []

for row in cleaner_app:
    price = row[5]
    
    if price == '0':
        cleanest_app.append(row)

for row in cleaner_play:
    price = row[7]
    
    if price == '0':
        cleanest_play.append(row)        
        
print("Apps Store free apps have {} entries.".format(len(cleanest_app)))
print("Play Store free apps have {} entries.".format(len(cleanest_play)))        

Apps Store free apps have 3222 entries.
Play Store free apps have 8864 entries.


**Part 2**

One of the column in our data that is useful to achive our goal is the `Genre` column. Perhaps an Social Media or Arts & Design app attracts a large amount of users. To test this thoery, I will created some frequency tables.

The two functions that will help us here are:
* `freq_table()` - Given a data set and index / colum, count the frequency of each item in the column and return the percentage of each item.
* `sorted_table()` - Give data set and index, retrieve the frequency percentage using frequency function, then sort and print each item from highest to lowest.

In [11]:
def freq_table(dataset, index):
    freq_dict = {}
    
    for row in dataset:
        column_data = row[index]
        
        if column_data in freq_dict:
            freq_dict[column_data] += 1
        else:
            freq_dict[column_data] = 1
            
    freq_percentage = {}
    
    for key in freq_dict:
        percentage = (freq_dict[key] / len(dataset)) * 100
        freq_percentage[key] = percentage
            
    return freq_percentage


def sorted_table(dataset, index):
    frequency_table = freq_table(dataset, index)
    
    table = sorted(frequency_table.items(), key=lambda items:items[1], reverse=True)
    
    for item in table:
        print(item[0] + ':', item[1])
        
    return table
        


**Part 3**

Apple App Store - Genre Column

It seems the genre **Games** is the clear winner our Apple Apps Store data set making up 58.16% of all total apps. The running up **Entertainment** at 7.88%, followed by **Photo & Video** at 4.96%. At the bottom, **Catalogs**, **Medical**, **Navigation** and **Book** each make up of less than 0.5% of total apps.

Although the **Games** genre has the most number of apps, it may not be true for the most number of users. Thus, based on only the data from types of apps below I am unable to recommend a profile for the type of iOS apps to create yet.

In [12]:
app_genre = sorted_table(cleanest_app, 12)

Games: 58.16263190564867
Entertainment: 7.883302296710118
Photo & Video: 4.9658597144630665
Education: 3.662321539416512
Social Networking: 3.2898820608317814
Shopping: 2.60707635009311
Utilities: 2.5139664804469275
Sports: 2.1415270018621975
Music: 2.0484171322160147
Health & Fitness: 2.0173805090006205
Productivity: 1.7380509000620732
Lifestyle: 1.5828677839851024
News: 1.3345747982619491
Travel: 1.2414649286157666
Finance: 1.1173184357541899
Weather: 0.8690254500310366
Food & Drink: 0.8069522036002483
Reference: 0.5586592178770949
Business: 0.5276225946617008
Book: 0.4345127250155183
Navigation: 0.186219739292365
Medical: 0.186219739292365
Catalogs: 0.12414649286157665


Google Play Store - Category Column

From the Google Play Store, **FAMILY** category has the most number of apps at 18.9%. Followed by **GAME** at 9.72%, and **TOOLS** at 8.46%. At the bottom we can find **BEAUTY**, **COMICS**, and **ART_AND_DESIGN** at less than 1%.

Although family friendly apps are dominating the Play Store, it is not as skewed like the App Store. Overall, the data is more spread out and balanced. The FAMILY category seems to be geared towards kids and some of those apps might fit into other categories such as GAMES, TOOLS, and LIFESTYLE. Thus, I am also unable to recommend an app profile to create for the Play Store based on this data alone.

In [13]:
play_category = sorted_table(cleanest_play, 1)

FAMILY: 18.907942238267147
GAME: 9.724729241877256
TOOLS: 8.461191335740072
BUSINESS: 4.591606498194946
LIFESTYLE: 3.9034296028880866
PRODUCTIVITY: 3.892148014440433
FINANCE: 3.7003610108303246
MEDICAL: 3.531137184115524
SPORTS: 3.395758122743682
PERSONALIZATION: 3.3167870036101084
COMMUNICATION: 3.2378158844765346
HEALTH_AND_FITNESS: 3.0798736462093865
PHOTOGRAPHY: 2.944494584837545
NEWS_AND_MAGAZINES: 2.7978339350180503
SOCIAL: 2.6624548736462095
TRAVEL_AND_LOCAL: 2.33528880866426
SHOPPING: 2.2450361010830324
BOOKS_AND_REFERENCE: 2.1435018050541514
DATING: 1.861462093862816
VIDEO_PLAYERS: 1.7937725631768955
MAPS_AND_NAVIGATION: 1.3989169675090252
FOOD_AND_DRINK: 1.2409747292418771
EDUCATION: 1.1620036101083033
ENTERTAINMENT: 0.9589350180505415
LIBRARIES_AND_DEMO: 0.9363718411552346
AUTO_AND_VEHICLES: 0.9250902527075812
HOUSE_AND_HOME: 0.8235559566787004
WEATHER: 0.8009927797833934
EVENTS: 0.7107400722021661
PARENTING: 0.6543321299638989
ART_AND_DESIGN: 0.6430505415162455
COMICS: 0.62

Google Play Store - Genre Column

The genres data in the Google Play Store data set contains a very high number of categories. This is much more granular than the category data. As we do not need to be so detailed, we can ignore the genre data from the Google data set and just use the Category data for comparison.

In [14]:
play_genre = sorted_table(cleanest_play, 9)

Tools: 8.449909747292418
Entertainment: 6.069494584837545
Education: 5.347472924187725
Business: 4.591606498194946
Lifestyle: 3.892148014440433
Productivity: 3.892148014440433
Finance: 3.7003610108303246
Medical: 3.531137184115524
Sports: 3.463447653429603
Personalization: 3.3167870036101084
Communication: 3.2378158844765346
Action: 3.1024368231046933
Health & Fitness: 3.0798736462093865
Photography: 2.944494584837545
News & Magazines: 2.7978339350180503
Social: 2.6624548736462095
Travel & Local: 2.3240072202166067
Shopping: 2.2450361010830324
Books & Reference: 2.1435018050541514
Simulation: 2.0419675090252705
Dating: 1.861462093862816
Arcade: 1.8501805054151623
Video Players & Editors: 1.7712093862815883
Casual: 1.7599277978339352
Maps & Navigation: 1.3989169675090252
Food & Drink: 1.2409747292418771
Puzzle: 1.128158844765343
Racing: 0.9927797833935018
Libraries & Demo: 0.9363718411552346
Role Playing: 0.9363718411552346
Auto & Vehicles: 0.9250902527075812
Strategy: 0.913808664259927

## Most Popular App by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` (column 12) app's data.

Below, we calculate the average number of user ratings per app genre on the App Store:


In [15]:
app_store_genre = freq_table(cleanest_app, 12) # Frequency dictionary for which the keys are the genres
genre_avg_table = {} # dictionary to store average number of ratings for each genre.

# Nested for loop that goes through the data set for each genre and count total number of ratings and find average.
for genre in app_store_genre:
    total = 0       # Total number of ratings for the current genre.
    len_genre = 0   # Total number of apps of current genre.
    
    for app in cleanest_app:
        genre_app = app[12] # Name of app in current iteration of data set.
        
        if genre == genre_app:
            total += float(app[6])
            len_genre += 1

    average = total / len_genre
    
    genre_avg_table[genre] = average

for k, v in sorted(genre_avg_table.items(), key=lambda items:items[1], reverse=True):
    print(k, ':', v)

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


We will look at individual genre more closely by creating a function that will tell us how many installs or number of ratings for all apps of a specific genre in our data set.

* `dataset` - list of list of a dataset.
* `genre` - string of the genre type (Games, Navigation, Sports, etc) to look for.
* `genre_column` - int column number where the genre is found in the inner list of the data set.
* `name_column` - int column number where the name of the app is found in the inner list of the data set.
* `count_column` - int coumn number where the number of rating or installs is fount in the inner list of the data set.

In [16]:
def genre_count(dataset, genre, genre_column, name_column, count_column):
    genre_count = {}
    
    for row in dataset:
        if row[genre_column] == genre:
            genre_count[row[name_column]] = int(row[count_column].replace('+', '').replace(',', ''))
            
    return genre_count

On average, **Navigation** apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [17]:
nav_app = genre_count(cleanest_app, 'Navigation', 12, 2, 6)

for k, v in sorted(nav_app.items(), key=lambda items: items[1], reverse=True):
    print(k, ':', f'{v:,}')

Waze - GPS Navigation, Maps & Real-time Traffic : 345,046
Google Maps - Navigation & Transit : 154,911
Geocaching® : 12,811
CoPilot GPS – Car Navigation & Offline Maps : 3,582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

**Reference** apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:


In [18]:
nav_app = genre_count(cleanest_app, 'Reference', 12, 2, 6)

for k, v in sorted(nav_app.items(), key=lambda items: items[1], reverse=True):
    print(k, ':', f'{v:,}')

Bible : 985,920
Dictionary.com Dictionary & Thesaurus : 200,047
Dictionary.com Dictionary & Thesaurus for iPad : 54,175
Google Translate : 26,786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18,418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17,588
Merriam-Webster Dictionary : 16,849
Night Sky : 12,122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8,535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4,693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1,497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
Jishokun-Japanese English Dictionary & Translator : 0
教えて!goo : 0


The same pattern applies to **social networking apps**, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

In [19]:
nav_app = genre_count(cleanest_app, 'Social Networking', 12, 2, 6)

for k, v in sorted(nav_app.items(), key=lambda items: items[1], reverse=True):
    print(k, ':', f'{v:,}')

Facebook : 2,974,676
Pinterest : 1,061,624
Skype for iPhone : 373,519
Messenger : 351,466
Tumblr : 334,293
WhatsApp Messenger : 287,589
Kik : 260,965
ooVoo – Free Video Call, Text and Voice : 177,501
TextNow - Unlimited Text + Calls : 164,963
Viber Messenger – Text & Call : 164,249
Followers - Social Analytics For Instagram : 112,778
MeetMe - Chat and Meet New People : 97,072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90,414
InsTrack for Instagram - Analytics Plus More : 85,535
Tango - Free Video Call, Voice and Chat : 75,412
LinkedIn : 71,856
Match™ - #1 Dating App. : 60,659
Skype for iPad : 60,163
POF - Best Dating App for Conversations : 52,642
Timehop : 49,510
Find My Family, Friends & iPhone - Life360 Locator : 43,877
Whisper - Share, Express, Meet : 39,819
Hangouts : 36,404
LINE PLAY - Your Avatar World : 34,677
WeChat : 34,584
Badoo - Meet New People, Chat, Socialize. : 34,428
Followers + for Instagram - Follower Analytics : 28,633
GroupMe : 28,260
Marco Polo Video Wal

**Summary of App Store**



However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

* Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

* Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

* Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

## Most Popular Apps by Genre on Google Play

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.).

One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

In [20]:
play_installs = sorted_table(cleanest_play, 5)

1,000,000+: 15.726534296028879
100,000+: 11.552346570397113
10,000,000+: 10.548285198555957
10,000+: 10.198555956678701
1,000+: 8.393501805054152
100+: 6.915613718411552
5,000,000+: 6.825361010830325
500,000+: 5.561823104693141
50,000+: 4.7721119133574
5,000+: 4.512635379061372
10+: 3.5424187725631766
500+: 3.2490974729241873
50,000,000+: 2.3014440433213
100,000,000+: 2.1322202166064983
50+: 1.917870036101083
5+: 0.78971119133574
1+: 0.5076714801444043
500,000,000+: 0.2707581227436823
1,000,000,000+: 0.22563176895306858
0+: 0.04512635379061372
0: 0.01128158844765343


To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [21]:
play_category = freq_table(cleanest_play, 1) # Although this gives us percentage, all we need is just the unique keys.

play_category_table = {}

for app in play_category:
    len_app = 0
    total_app_installs = 0
    
    for row in cleanest_play:
        app_name = row[1]
        
        if app == app_name:
            len_app += 1
            installs = float(row[5].replace('+', '').replace(',', ''))
            total_app_installs += installs
            
    average = total_app_installs / len_app
    play_category_table[app] = average
    
for k, v in sorted(play_category_table.items(), key=lambda items: items[1], reverse=True):
    print(k, ':', f'{v:,}')

COMMUNICATION : 38,456,119.167247385
VIDEO_PLAYERS : 24,727,872.452830188
SOCIAL : 23,253,652.127118643
PHOTOGRAPHY : 17,840,110.40229885
PRODUCTIVITY : 16,787,331.344927534
GAME : 15,588,015.603248259
TRAVEL_AND_LOCAL : 13,984,077.710144928
ENTERTAINMENT : 11,640,705.88235294
TOOLS : 10,801,391.298666667
NEWS_AND_MAGAZINES : 9,549,178.467741935
BOOKS_AND_REFERENCE : 8,767,811.894736841
SHOPPING : 7,036,877.311557789
PERSONALIZATION : 5,201,482.6122448975
WEATHER : 5,074,486.197183099
HEALTH_AND_FITNESS : 4,188,821.9853479853
MAPS_AND_NAVIGATION : 4,056,941.7741935486
FAMILY : 3,695,641.8198090694
SPORTS : 3,638,640.1428571427
ART_AND_DESIGN : 1,986,335.0877192982
FOOD_AND_DRINK : 1,924,897.7363636363
EDUCATION : 1,833,495.145631068
BUSINESS : 1,712,290.1474201474
LIFESTYLE : 1,437,816.2687861272
FINANCE : 1,387,692.475609756
HOUSE_AND_HOME : 1,331,540.5616438356
DATING : 854,028.8303030303
COMICS : 817,657.2727272727
AUTO_AND_VEHICLES : 647,317.8170731707
LIBRARIES_AND_DEMO : 638,503.

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [22]:
com_app = genre_count(cleanest_play, 'COMMUNICATION', 1, 0, 5)

for k, v in sorted(com_app.items(), key=lambda items: items[1], reverse=True):
    if v >= 100_000_000:
        print(k, ':', f'{v:,}')


WhatsApp Messenger : 1,000,000,000
Messenger – Text and Video Chat for Free : 1,000,000,000
Skype - free IM & video calls : 1,000,000,000
Google Chrome: Fast & Secure : 1,000,000,000
Gmail : 1,000,000,000
Hangouts : 1,000,000,000
Google Duo - High Quality Video Calls : 500,000,000
imo free video calls and chat : 500,000,000
LINE: Free Calls & Messages : 500,000,000
UC Browser - Fast Download Private & Secure : 500,000,000
Viber Messenger : 500,000,000
imo beta free calls and text : 100,000,000
Android Messages : 100,000,000
Who : 100,000,000
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000
Firefox Browser fast & private : 100,000,000
Messenger Lite: Free Calls & Messages : 100,000,000
Kik : 100,000,000
KakaoTalk: Free Calls & Text : 100,000,000
Opera Mini - fast web browser : 100,000,000
Opera Browser: Fast and Secure : 100,000,000
Telegram : 100,000,000
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000
WeC

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [23]:
under_100m_total = 0
under_100m_num_app = 0

for app, installs in com_app.items():
    if installs < 100_000_000:
        under_100m_total += installs
        under_100m_num_app += 1

average = under_100m_total / under_100m_num_app

print('Average installs per app for COMMUNICATION apps that are under 100 millions installs:', average)


Average installs per app for COMMUNICATION apps that are under 100 millions installs: 3603485.3884615386


**Further Analysis**

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:


In [24]:
com_app = genre_count(cleanest_play, 'BOOKS_AND_REFERENCE', 1, 0, 5)

for k, v in sorted(com_app.items(), key=lambda items: items[1], reverse=True):
    print(k, ':', f'{v:,}')

Google Play Books : 1,000,000,000
Bible : 100,000,000
Amazon Kindle : 100,000,000
Wattpad 📖 Free Books : 100,000,000
Audiobooks from Audible : 100,000,000
Wikipedia : 10,000,000
Cool Reader : 10,000,000
FBReader: Favorite Book Reader : 10,000,000
HTC Help : 10,000,000
Moon+ Reader : 10,000,000
Aldiko Book Reader : 10,000,000
Al-Quran (Free) : 10,000,000
Al Quran Indonesia : 10,000,000
Al'Quran Bahasa Indonesia : 10,000,000
Quran for Android : 10,000,000
Dictionary.com: Find Definitions for English Words : 10,000,000
English Dictionary - Offline : 10,000,000
NOOK: Read eBooks & Magazines : 10,000,000
Dictionary : 10,000,000
Spanish English Translator : 10,000,000
Dictionary - Merriam-Webster : 10,000,000
JW Library : 10,000,000
Oxford Dictionary of English : Free : 10,000,000
English Hindi Dictionary : 10,000,000
AlReader -any text book reader : 5,000,000
Ebook Reader : 5,000,000
Read books online : 5,000,000
Ancestry : 5,000,000
Dictionary - WordWeb : 5,000,000
50000 Free eBooks & Free

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [29]:
for k, v in sorted(com_app.items(), key=lambda items: items[1], reverse=True):
    if v >= 100_000_000:
        print(k, ':', f'{v:,}')

Google Play Books : 1,000,000,000
Bible : 100,000,000
Amazon Kindle : 100,000,000
Wattpad 📖 Free Books : 100,000,000
Audiobooks from Audible : 100,000,000


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [30]:
for k, v in sorted(com_app.items(), key=lambda items: items[1], reverse=True):
    if 1_000_000 <= v <= 50_000_000:
        print(k, ':', f'{v:,}')

Wikipedia : 10,000,000
Cool Reader : 10,000,000
FBReader: Favorite Book Reader : 10,000,000
HTC Help : 10,000,000
Moon+ Reader : 10,000,000
Aldiko Book Reader : 10,000,000
Al-Quran (Free) : 10,000,000
Al Quran Indonesia : 10,000,000
Al'Quran Bahasa Indonesia : 10,000,000
Quran for Android : 10,000,000
Dictionary.com: Find Definitions for English Words : 10,000,000
English Dictionary - Offline : 10,000,000
NOOK: Read eBooks & Magazines : 10,000,000
Dictionary : 10,000,000
Spanish English Translator : 10,000,000
Dictionary - Merriam-Webster : 10,000,000
JW Library : 10,000,000
Oxford Dictionary of English : Free : 10,000,000
English Hindi Dictionary : 10,000,000
AlReader -any text book reader : 5,000,000
Ebook Reader : 5,000,000
Read books online : 5,000,000
Ancestry : 5,000,000
Dictionary - WordWeb : 5,000,000
50000 Free eBooks & Free AudioBooks : 5,000,000
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000
Bible KJV : 5,000,000
English to Hindi Dictionary : 5,000,000
Book store 

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.